### ISSUE TITLE
About the global null-text Inversion

### ISSUE THREAD

**Issue by RayleighChen** [google/prompt-to-prompt#23]
Excellent job!
I have some questions about Global null-text Inversion. I attempted to implement this algorithm, but I can't do a good reconstruction of the final optimized null-text embedding even if I change the number of optimization steps from 7500 to 10000. Could you kindly release the codes for this algorithm? Or is there any code snippet. Thanks!

```python
def global_null_optimization(self, latents, num_steps, epsilon):
    uncond_embeddings, cond_embeddings = self.context.chunk(2)

    bar = tqdm(total=num_steps * NUM_DDIM_STEPS)

    for j in range(num_steps):

        latent_cur = latents[-1]
        for i in range(NUM_DDIM_STEPS):
            uncond_embeddings = uncond_embeddings.clone().detach()
            uncond_embeddings.requires_grad = True
            optimizer = Adam([uncond_embeddings], lr=1e-2 * (1. - i / 100.))
            latent_prev = latents[len(latents) - i - 2]
            t = self.model.scheduler.timesteps[i]
            with torch.no_grad():
                noise_pred_cond = self.get_noise_pred_single(latent_cur, t, cond_embeddings)

            noise_pred_uncond = self.get_noise_pred_single(latent_cur, t, uncond_embeddings)
            noise_pred = noise_pred_uncond + GUIDANCE_SCALE * (noise_pred_cond - noise_pred_uncond)
            latents_prev_rec = self.prev_step(noise_pred, t, latent_cur)
            loss = nnf.mse_loss(latents_prev_rec, latent_prev)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            loss_item = loss.item()
            bar.update()
            if loss_item < epsilon + i * 2e-5:
                break

            with torch.no_grad():
                context = torch.cat([uncond_embeddings, cond_embeddings])
                latent_cur = self.get_noise_pred(latent_cur, t, False, context)

        for i in range(i + 1, NUM_DDIM_STEPS):
            bar.update()

    bar.close()
    return uncond_embeddings.detach()
```

**Comment by Yolice** [comment]
Hello, I also encountered a similar problem. Can you share your global implementation code? Thanks

**Comment by mshu1** [comment]
Hi Rayleigh! I was trying to implement the global null-text inversion as well and ran into the same issue (also unstable loss function) How did you solve the problem? Do you mind sharing what you implemented? Thanks!

**Comment by Open-Debin** [comment]
@RayleighChen @mshu1 Hello, I am very interested in the global null-text inversion. May I have your code. I am very thank you if you could help to share this code. Thank you very much!
